lib/service/storage/virtualIndex.ts

```diff
@@ -0,0 +1,200 @@
+import { Service } from "../service";
+
+export class VirtualIndex extends Service {
+  public static createEvent = "core:storage:virtual-index:cache:create";
+  public static deleteEvent = "core:storage:virtual-index:cache:delete";
+
+  public static createVirtualIndexEvent =
+    "core:storage:virtual-index:create:after";
+
+  public static deleteVirtualIndexEvent =
+    "core:storage:virtual-index:delete:after";
+
+  constructor() {
+    super("VirtualIndex", global.kuzzle.config.services.storageEngine);
+  }
+
+  public virtualIndexMap: Map<string, string> = new Map(); //Key : virtual index, value : real index
+
+  async init(): Promise<void> {
+    await this.buildCollection();
+    await this.initVirtualIndexList();
+
+    global.kuzzle.onAsk(VirtualIndex.createEvent, (physical, virtual) =>
+      this.addInSoftIndexMap(physical, virtual)
+    );
+
+    global.kuzzle.onAsk(VirtualIndex.deleteEvent, (info) =>
+      this.removeInSoftIndexMap(info)
+    );
+  }
+
+  addInSoftIndexMap(physical, virtual) {
+    this.virtualIndexMap.set(virtual, physical);
+  }
+
+  removeInSoftIndexMap(notification) {
+    this.virtualIndexMap.delete(notification);
+  }
+
+  getRealIndex(name: string): string {
+    if (this.virtualIndexMap.has(name)) {
+      return this.virtualIndexMap.get(name);
+    }
+    return name;
+  }
+
+  isVirtual(name: string): boolean {
+    return this.virtualIndexMap.has(name);
+  }
+
+  getId(index: string, id: string): string {
+    if (this.isVirtual(index)) {
+      return index + id;
+    }
+    return id;
+  }
+
+  randomString(size = 20) {
+    const buf = Buffer.alloc(size);
+    for (let i = 0; i < 20; i++) {
+      buf.writeUInt8((Math.random() * 255) & 255, i);
+    }
+    return buf.toString("base64").slice(0, size);
+  }
+
+  getVirtualId(index: string, id: string): string {
+    if (this.isVirtual(index) && id.startsWith(index)) {
+      return id.substring(index.length, id.length);
+    }
+    return id;
+  }
+
+  public async createVirtualIndex(virtualIndex: string, index: string) {
+    global.kuzzle.emit(VirtualIndex.createVirtualIndexEvent, {
+      physical: index,
+      virtual: virtualIndex,
+    });
+
+    this.virtualIndexMap.set(virtualIndex, index);
+    await global.kuzzle.ask(
+      "core:storage:private:document:create",
+      "virtual-indexes",
+      "list",
+      { physical: index, virtual: virtualIndex },
+      { id: index + virtualIndex }
+    );
+  }
+
+  async removeVirtualIndex(index: string) {
+    const realIndex = this.virtualIndexMap.get(index);
+    global.kuzzle.emit(VirtualIndex.deleteVirtualIndexEvent, {
+      virtual: index,
+    });
+    this.virtualIndexMap.delete(index);
+    const id = realIndex + index;
+    await global.kuzzle.ask(
+      "core:storage:private:document:delete",
+      "virtual-indexes",
+      "list",
+      id
+    );
+  }
+
+  async initVirtualIndexList() {
+    if (this.virtualIndexMap.size === 0) {
```

Comment:
I believe that if you are returning when it's empty then when Kuzzle starts you are not going to fill the `virtualIndexMap`

Reply:
Done